### PR TITLE
feat(linter/node): implement no-new-require

### PR DESCRIPTION
**Pull request by jelly**

### CODE REVIEW

**Review by camc314 — APPROVED** [review]

File: crates/oxc_linter/src/rules/node/no_new_require.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        "var appHeader = require('app-header')",
+        "var AppHeader = new (require('app-header'))",
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/node/no_new_require.rs:62]: Why should this be allowed?

File: crates/oxc_linter/src/rules/node/no_new_require.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        "var AppHeader = new (require('headers').appHeader)",
+        "var AppHeader = require('app-header'); var appHeader = new AppHeader();",
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/node/no_new_require.rs:64]: Can you please add a pass case for `new require('foo')()`

File: crates/oxc_linter/src/rules/node/no_new_require.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_new_require(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Unexpected use of new with require")
+        .with_label(span)
+        .with_help("Initialise the constructor separate from the import statement")
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/node/no_new_require.rs:11]: This message is a little confusing

File: crates/oxc_linter/src/rules/node/no_new_require.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+pub struct NoNewRequire;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Warn about calling `new` on `require`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// The `require` function is used to include modules and might return a constructor. As this
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/node/no_new_require.rs:24]: We should mention that `require` is a non-newable function, so if they're not using a custom function also called `require` then they'll get a runtime error.

File: crates/oxc_linter/src/rules/node/no_new_require.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+fn no_new_require(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Unexpected use of new with require")
+        .with_label(span)
+        .with_help("Initialise the constructor separate from the import statement")
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoNewRequire;
+
+declare_oxc_lint!(
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/node/no_new_require.rs:17]: Prefer curly braces for `declare_oxc_lint!`

File: crates/oxc_linter/src/rules/node/no_new_require.rs (hunk 1 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/node/no_new_require.rs:62]: That's what the original rule allows as well, if we want to be more strict I would be happy to implement that behaviour.

File: crates/oxc_linter/src/rules/node/no_new_require.rs (hunk 4 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/node/no_new_require.rs:24]: test.js
```javascript
const x = new require("./lol.js");
console.log(x);
```

foo.js
```javascript
class Foo {
};

module.exports = Foo;
```

This does not throw a runtime error with nodejs 22.

Comment (jelly) [reply crates/oxc_linter/src/rules/node/no_new_require.rs:24]: Similar example here https://nodejs.org/docs/v20.17.0/api/modules.html#modules-commonjs-modules

Comment (DonIsaac) [reply crates/oxc_linter/src/rules/node/no_new_require.rs:24]: You make a good point. This is pathological, but valid

File: crates/oxc_linter/src/rules/node/no_new_require.rs (hunk 1 above)
Comment (nrayburn-tech) [reply crates/oxc_linter/src/rules/node/no_new_require.rs:62]: Parenthesis are optional when constructing an instance that doesn't take any arguments.  Confusing but valid.

**Pull request merged**